File: traffic_engine.py
import random
import time
from collections import deque

class TrafficSimulator:
    def __init__(self, num_lanes=4, max_capacity=20.0):
        self.num_lanes = num_lanes
        self.max_capacity = max_capacity
        self.lanes = {i: {
            'occupancy': 0.0, 
            'history': deque([0.0], maxlen=10),
            'anti_gravity': False, 
            'signal': 'GREEN'
        } for i in range(1, num_lanes + 1)}
        
        self.vehicle_types = {
            'Scooter': 0.2,
            'Auto': 0.5,
            'Car': 1.0,
            'Bus': 2.5
        }
        
        self.total_counts = {v: 0 for v in self.vehicle_types.keys()}
        self.stress_lanes = {i: 0 for i in range(1, num_lanes + 1)} 
        self.hidden_scooters = {i: 0 for i in range(1, num_lanes + 1)} 
        self.efficiency_history = deque([100.0], maxlen=10)
        
        self.logs = []
        self.jam_threshold = 0.25 
        self.min_occupancy_for_jam = 2.0

    def update(self):
        current_time = time.strftime("%H:%M:%S")
        total_occupancy = 0
        total_departure = 0
        
        for lane_id, data in self.lanes.items():
            new_vehicle_mass = 0
            v_type = None
            
            if self.stress_lanes[lane_id] > 0:
                v_type = random.choice(['Bus', 'Car'])
                new_vehicle_mass = self.vehicle_types[v_type]
                self.total_counts[v_type] += 1
                self.stress_lanes[lane_id] -= 1
            else:
                if random.random() > 0.4:
                    v_type = random.choice(list(self.vehicle_types.keys()))
                    new_vehicle_mass = self.vehicle_types[v_type]
                    self.total_counts[v_type] += 1
            
            if v_type in ['Bus', 'Car', 'Auto'] and random.random() > 0.7:
                spillover = new_vehicle_mass * 0.4
                new_vehicle_mass -= spillover
                adj_lane = lane_id + 1 if lane_id < self.num_lanes else lane_id - 1
                self.lanes[adj_lane]['occupancy'] = min(self.max_capacity, self.lanes[adj_lane]['occupancy'] + spillover)
                
            if v_type == 'Bus' and random.random() > 0.5:
                self.hidden_scooters[lane_id] += random.randint(1, 3)
            
            departure_rate = 0.3 if not data['anti_gravity'] else 1.2 
            if data['occupancy'] > 0.8 * self.max_capacity:
                departure_rate *= 0.5

            actual_departure = min(data['occupancy'], departure_rate)
            total_departure += actual_departure

            data['occupancy'] = max(0, min(self.max_capacity, round(data['occupancy'] + new_vehicle_mass - actual_departure, 2)))
            total_occupancy += data['occupancy']
            
            self._run_4d_radar(lane_id, data, current_time)
            
            self._predict_jam(lane_id, data, current_time)
            
            data['history'].append(round((data['occupancy'] / self.max_capacity) * 100, 2))

        global_capacity = self.num_lanes * self.max_capacity
        efficiency = max(0, min(100, 100 - (total_occupancy / global_capacity * 100)))
        self.efficiency_history.append(round(efficiency, 1))

    def _predict_jam(self, lane_id, data, current_time):
        if len(data['history']) < 5:
            return

        recent_history = list(data['history'])
        growth = (recent_history[-1] - recent_history[-4]) / (recent_history[-4] if recent_history[-4] > 0 else 1)
        
        if (growth > self.jam_threshold or data['occupancy'] > 0.8 * self.max_capacity) and data['occupancy'] > self.min_occupancy_for_jam:
            if not data['anti_gravity']:
                data['anti_gravity'] = True
                self.logs.append(f"[{current_time}] Lane {lane_id}: Critical Density! Activating Smooth Flow.")
        elif data['occupancy'] < 0.25 * self.max_capacity:
            if data['anti_gravity']:
                data['anti_gravity'] = False
                self.logs.append(f"[{current_time}] Lane {lane_id}: Flow Restored. Disengaging Smooth Flow.")

    def trigger_stress_test(self, lane_id):
        if lane_id in self.stress_lanes:
            self.stress_lanes[lane_id] = 5 
            self.logs.append(f"[{time.strftime('%H:%M:%S')}] User Action: Manual Flood Triggered for Lane {lane_id}.")
            return True
        return False

    def trigger_ambulance(self):
        densest_lane = max(self.lanes.keys(), key=lambda l: self.lanes[l]['occupancy'])
        self.lanes[densest_lane]['occupancy'] = 0.0
        self.lanes[densest_lane]['anti_gravity'] = False
        self.logs.append(f"[{time.strftime('%H:%M:%S')}] 🚨 EMERGENCY GREEN WAVE: Ambulance detected. Lane {densest_lane} instantly cleared.")
        return densest_lane

    def _run_4d_radar(self, lane_id, data, current_time):
        if self.hidden_scooters[lane_id] > 0:
            hidden_mass = self.hidden_scooters[lane_id] * self.vehicle_types['Scooter']
            effective_mass = data['occupancy'] + hidden_mass
            if effective_mass > data['occupancy']:
                data['occupancy'] = min(self.max_capacity, round(effective_mass, 2))
                self.logs.append(f"[{current_time}] 📡 Lane {lane_id}: 4D Radar detected {self.hidden_scooters[lane_id]} hidden scooter(s). Re-evaluating density.")
            self.hidden_scooters[lane_id] = max(0, self.hidden_scooters[lane_id] - 1) 

    def get_status(self):
        status = {
            'lanes': {lane_id: {
                'occupancy': round(data['occupancy'], 2),
                'density_pct': data['history'][-1] if data['history'] else 0,
                'anti_gravity': data['anti_gravity'],
                'history': list(data['history'])
            } for lane_id, data in self.lanes.items()},
            'logs': self.logs[-15:],
            'stats': {
                'vehicle_counts': self.total_counts,
                'efficiency': self.efficiency_history[-1] if self.efficiency_history else 100.0,
                'is_stressing': self.stress_lanes
            }
        }
        return status

File: Traffic_Engine.js
class Traffic_Simulator {
    constructor(Num_Lanes = 4, Max_Capacity = 20.0) {
        this.Num_Lanes = Num_Lanes;
        this.Max_Capacity = Max_Capacity;
        this.Lanes = {};
        for (let I = 1; I <= Num_Lanes; I++) {
            this.Lanes[I] = {
                Occupancy: 0.0,
                History: [0.0],
                Anti_Gravity: false,
                Signal: 'GREEN'
            };
        }

        this.Vehicle_Types = {
            'Scooter': 0.2,
            'Auto': 0.5,
            'Car': 1.0,
            'Bus': 2.5
        };

        this.Total_Counts = {};
        for (const V in this.Vehicle_Types) {
            this.Total_Counts[V] = 0;
        }

        this.Stress_Lanes = {};
        this.Hidden_Scooters = {};
        for (let I = 1; I <= Num_Lanes; I++) {
            this.Stress_Lanes[I] = 0;
            this.Hidden_Scooters[I] = 0;
        }

        this.Efficiency_History = [100.0];
        this.Logs = [];
        this.Jam_Threshold = 0.25;
        this.Min_Occupancy_For_Jam = 2.0;
    }

    Update() {
        const Now = new Date();
        const Current_Time = Now.toTimeString().split(' ')[0];
        let Total_Occupancy = 0;
        let Total_Departure = 0;

        for (let Lane_Id = 1; Lane_Id <= this.Num_Lanes; Lane_Id++) {
            const Data = this.Lanes[Lane_Id];
            let New_Vehicle_Mass = 0;
            let V_Type = null;

            if (this.Stress_Lanes[Lane_Id] > 0) {
                const Choices = ['Bus', 'Car'];
                V_Type = Choices[Math.floor(Math.random() * Choices.length)];
                New_Vehicle_Mass = this.Vehicle_Types[V_Type];
                this.Total_Counts[V_Type]++;
                this.Stress_Lanes[Lane_Id]--;
            } else {
                if (Math.random() > 0.4) {
                    const Keys = Object.keys(this.Vehicle_Types);
                    V_Type = Keys[Math.floor(Math.random() * Keys.length)];
                    New_Vehicle_Mass = this.Vehicle_Types[V_Type];
                    this.Total_Counts[V_Type]++;
                }
            }

            if (['Bus', 'Car', 'Auto'].includes(V_Type) && Math.random() > 0.7) {
                const Spillover = New_Vehicle_Mass * 0.4;
                New_Vehicle_Mass -= Spillover;
                const Adj_Lane = Lane_Id < this.Num_Lanes ? Lane_Id + 1 : Lane_Id - 1;
                this.Lanes[Adj_Lane].Occupancy = Math.min(this.Max_Capacity, this.Lanes[Adj_Lane].Occupancy + Spillover);
            }

            if (V_Type === 'Bus' && Math.random() > 0.5) {
                this.Hidden_Scooters[Lane_Id] += Math.floor(Math.random() * 3) + 1;
            }

            let Departure_Rate = !Data.Anti_Gravity ? 0.3 : 1.2;
            if (Data.Occupancy > 0.8 * this.Max_Capacity) {
                Departure_Rate *= 0.5;
            }

            const Actual_Departure = Math.min(Data.Occupancy, Departure_Rate);
            Total_Departure += Actual_Departure;

            Data.Occupancy = Math.max(0, Math.min(this.Max_Capacity, Math.round((Data.Occupancy + New_Vehicle_Mass - Actual_Departure) * 100) / 100));
            Total_Occupancy += Data.Occupancy;

            this.Run_4d_Radar(Lane_Id, Data, Current_Time);
            this.Predict_Jam(Lane_Id, Data, Current_Time);

            Data.History.push(Math.round((Data.Occupancy / this.Max_Capacity) * 10000) / 100);
            if (Data.History.length > 10) {
                Data.History.shift();
            }
        }

        const Global_Capacity = this.Num_Lanes * this.Max_Capacity;
        const Efficiency = Math.max(0, Math.min(100, 100 - (Total_Occupancy / Global_Capacity * 100)));
        this.Efficiency_History.push(Math.round(Efficiency * 10) / 10);
        if (this.Efficiency_History.length > 10) {
            this.Efficiency_History.shift();
        }
    }

    Predict_Jam(Lane_Id, Data, Current_Time) {
        if (Data.History.length < 5) return;

        const Recent = Data.History;
        const Growth = (Recent[Recent.length - 1] - Recent[Recent.length - 4]) / (Recent[Recent.length - 4] > 0 ? Recent[Recent.length - 4] : 1);

        if ((Growth > this.Jam_Threshold || Data.Occupancy > 0.8 * this.Max_Capacity) && Data.Occupancy > this.Min_Occupancy_For_Jam) {
            if (!Data.Anti_Gravity) {
                Data.Anti_Gravity = true;
                this.Logs.push(`[${Current_Time}] Lane ${Lane_Id}: Critical Density! Activating Smooth Flow.`);
            }
        } else if (Data.Occupancy < 0.25 * this.Max_Capacity) {
            if (Data.Anti_Gravity) {
                Data.Anti_Gravity = false;
                this.Logs.push(`[${Current_Time}] Lane ${Lane_Id}: Flow Restored. Disengaging Smooth Flow.`);
            }
        }
    }

    Trigger_Stress_Test(Lane_Id) {
        if (Lane_Id in this.Stress_Lanes) {
            this.Stress_Lanes[Lane_Id] = 5;
            const Now = new Date();
            this.Logs.push(`[${Now.toTimeString().split(' ')[0]}] User Action: Manual Flood Triggered for Lane ${Lane_Id}.`);
            return true;
        }
        return false;
    }

    Trigger_Ambulance() {
        let Densest_Lane = '1';
        let Max_Occupancy = -1;
        for (const L in this.Lanes) {
            if (this.Lanes[L].Occupancy > Max_Occupancy) {
                Max_Occupancy = this.Lanes[L].Occupancy;
                Densest_Lane = L;
            }
        }
        this.Lanes[Densest_Lane].Occupancy = 0.0;
        this.Lanes[Densest_Lane].Anti_Gravity = false;
        const Now = new Date();
        this.Logs.push(`[${Now.toTimeString().split(' ')[0]}] 🚨 EMERGENCY GREEN WAVE: Ambulance detected. Lane ${Densest_Lane} instantly cleared.`);
        return Densest_Lane;
    }

    Run_4d_Radar(Lane_Id, Data, Current_Time) {
        if (this.Hidden_Scooters[Lane_Id] > 0) {
            const Hidden_Mass = this.Hidden_Scooters[Lane_Id] * this.Vehicle_Types['Scooter'];
            const Effective_Mass = Data.Occupancy + Hidden_Mass;
            if (Effective_Mass > Data.Occupancy) {
                Data.Occupancy = Math.min(this.Max_Capacity, Math.round(Effective_Mass * 100) / 100);
                this.Logs.push(`[${Current_Time}] 📡 Lane ${Lane_Id}: 4D Radar detected ${this.Hidden_Scooters[Lane_Id]} hidden scooter(s). Re-evaluating density.`);
            }
            this.Hidden_Scooters[Lane_Id] = Math.max(0, this.Hidden_Scooters[Lane_Id] - 1);
        }
    }

    Get_Status() {
        const Status = {
            lanes: {},
            logs: this.Logs.slice(-15),
            stats: {
                vehicle_counts: this.Total_Counts,
                efficiency: this.Efficiency_History[this.Efficiency_History.length - 1] || 100.0,
                is_stressing: this.Stress_Lanes
            }
        };

        for (const L in this.Lanes) {
            const Data = this.Lanes[L];
            Status.lanes[L] = {
                occupancy: Math.round(Data.Occupancy * 100) / 100,
                density_pct: Data.History.length > 0 ? Data.History[Data.History.length - 1] : 0,
                anti_gravity: Data.Anti_Gravity,
                history: [...Data.History]
            };
        }
        return Status;
    }
}

module.exports = Traffic_Simulator;
